source/microbit/microbitcompass.cpp

```diff
@@ -31,93 +31,227 @@ extern "C" {
+    int8_t padding[1024];
+    compasss_calibration calibration;
+} persistent_compasss_data;
+
+static const persistent_compasss_data  __attribute__((aligned(1024))) data = {
+    0
+};
+
+static inline volatile compasss_calibration *get_calibration(void) {
+    return (volatile compasss_calibration *)&data.calibration;
+}
 
 typedef struct _microbit_compass_obj_t {
     mp_obj_base_t base;
     MicroBitCompass *compass;
 } microbit_compass_obj_t;
 
 mp_obj_t microbit_compass_is_calibrated(mp_obj_t self_in) {
-    microbit_compass_obj_t *self = (microbit_compass_obj_t*)self_in;
-    return mp_obj_new_bool(self->compass->isCalibrated());
+    (void)self_in;
+    return mp_obj_new_bool(get_calibration()->calibrated);
 }
 MP_DEFINE_CONST_FUN_OBJ_1(microbit_compass_is_calibrated_obj, microbit_compass_is_calibrated);
 
+extern mp_obj_t microbit_accelerometer_get_x(mp_obj_t self_in);
+extern mp_obj_t microbit_accelerometer_get_y(mp_obj_t self_in);
+
+static void update(microbit_compass_obj_t *self) {
+    /* The only time it is possible for compass_updating to be true here
+     * is if this is called in an interrupt when it is already updating in
+     * the main execution thread. This is extremely unlikely, so we just
+     * accept that a slightly out-of-date result will be returned
+     */
+    if (!compass_up_to_date && !compass_updating) {
+        compass_updating = true;
+        self->compass->idleTick();
+        compass_updating = false;
+        compass_up_to_date = true;
+    }
+}
+
+static void microbit_compass_save_calibration(bool calibrated, int x, int y, int z) {
+    persistent_erase_page((const void *)&data);
+    compasss_calibration calibration;
+    calibration.calibrated = calibrated;
+    calibration.x = x;
+    calibration.y = y;
+    calibration.z = z;
+    persistent_write((const void *)&data, &calibration, sizeof(compasss_calibration));
+}
+
+#define PERIMETER_POINTS 16
+
+struct Point
+{
+    uint8_t x;
+    uint8_t y;
+};
+
+static const Point perimeter[PERIMETER_POINTS] = {{0,0}, {1,0}, {2,0}, {3,0},
+                                            {4,0}, {4,1}, {4,2}, {4,3},
+                                            {4,4}, {3,4}, {2,4}, {1,4},
+                                            {0,4}, {0,3}, {0,2}, {0,1}};
+
 mp_obj_t microbit_compass_calibrate(mp_obj_t self_in) {
-    // Calibration requires to pass control over to the DAL so it
-    // can use the display to collect samples for the calibration.
-    // It will do the calibration and then return here.
     microbit_compass_obj_t *self = (microbit_compass_obj_t*)self_in;
-    ticker_stop();
-    uBit.systemTicker.attach_us(&uBit, &MicroBit::systemTick, MICROBIT_DEFAULT_TICK_PERIOD * 1000);
-    uBit.display.enable();
-    self->compass->calibrate();
-    uBit.display.disable();
-    uBit.systemTicker.detach();
-    ticker_start();
+    microbit_display_scroll(&microbit_display_obj, "Keep dot pointed ahead. Press button for each point.", false);
```

Comment:
Could this msg be shortened so it doesn't take so long to read, eg "Keep dot ahead, press button for next"?